scm/github/github_client.go

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptrace"
+	"net/url"
+	"strings"
+
+	"github.com/google/go-github/v65/github"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"golang.org/x/oauth2"
+
+	api "github.com/go-vela/server/api/types"
+	"github.com/go-vela/server/constants"
+)
+
+// newOAuthTokenClient returns the GitHub OAuth client.
+func (c *client) newOAuthTokenClient(ctx context.Context, token string) *github.Client {
+	// create the token object for the client
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+
+	// create the OAuth client
+	tc := oauth2.NewClient(ctx, ts)
+
+	if c.Tracing.Config.EnableTracing {
+		tc.Transport = otelhttp.NewTransport(
+			tc.Transport,
+			otelhttp.WithClientTrace(func(ctx context.Context) *httptrace.ClientTrace {
+				return otelhttptrace.NewClientTrace(ctx, otelhttptrace.WithoutSubSpans())
+			}),
+		)
+	}
+
+	// create the GitHub client from the OAuth client
+	github := github.NewClient(tc)
+
+	// ensure the proper URL is set in the GitHub client
+	github.BaseURL, _ = url.Parse(c.config.API)
+
+	return github
+}
+
+// newGithubAppClient returns the GitHub App client for authenticating as the GitHub App itself using the RoundTripper.
+func (c *client) newGithubAppClient() (*github.Client, error) {
+	if c.AppsTransport == nil {
+		return nil, errors.New("unable to create github app client: no AppsTransport configured")
+	}
+
+	// create a github client based off the existing GitHub App configuration
+	client, err := github.NewClient(
+		&http.Client{
+			Transport: c.AppsTransport,
+		}).
+		WithEnterpriseURLs(c.config.API, c.config.API)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// newGithubAppInstallationRepoToken returns the GitHub App installation token for a particular repo with granular permissions.
+func (c *client) newGithubAppInstallationRepoToken(ctx context.Context, r *api.Repo, repos []string, permissions *github.InstallationPermissions) (*github.InstallationToken, int64, error) {
+	// create a github client based off the existing GitHub App configuration
+	client, err := c.newGithubAppClient()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	opts := &github.InstallationTokenOptions{
+		Repositories: repos,
+		Permissions:  permissions,
+	}
+
+	id := r.GetInstallID()
+
+	// if repo has an install ID, use it to create an installation token
+	if id == 0 {
```

Comment:
nit: comment is a little confusing if we're checking if value is `0` right after. i think we're trying to get the id from installations if we DON'T have a installation ID. but i could be wrong 🫣

Reply:
good callout, thats confusing. i updated it as follows
> // if the source scm repo has an install ID but the Vela db record does not
> // then use the source repo to create an installation token